### PR TITLE
[V1] Get input tokens from scheduler

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR changes the scheduler and model runner so that the model runner gets the input token IDs from the scheduler. This change is especially useful when the token IDs are not generated by the model runner (e.g., non-last ranks in PP).

### CODE REVIEW

**Comment by WoosukKwon** [comment]
cc @comaniac This PR seems to work correctly when using TP (or single GPU), but PP still generates gibberish outputs.

**Review by comaniac — APPROVED** [review]
Awesome! I'll test with PP later.

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -341,12 +351,21 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> bool:
...
+                start_token_index = num_computed_tokens
+                end_token_index = num_computed_tokens + num_new_tokens
+                self.input_batch.token_ids_cpu[
+                    req_index,
+                    start_token_index:end_token_index] = new_token_ids
+                self.input_batch.num_tokens[req_index] += num_new_tokens
+
+        # Check if the batch has changed. If not, we can skip copying the
+        # sampling metadata from CPU to GPU.
         batch_changed = len(removed_req_indices) > 0 or len(req_ids_to_add) > 0
```
Comment (comaniac) [inline vllm/v1/worker/gpu_model_runner.py:369]: Just to make sure. Is this still correct? Even we have the same batch of requests, we may still need to copy the new tokens?

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -856,34 +875,21 @@ def execute_model(
         # TODO(woosuk): The following loop can be slow since it iterates over
         # the requests one by one. Optimize.
         num_reqs = self.input_batch.num_reqs
-        request_seq_lens: List[Tuple[int, CachedRequestState, int]] = []
-        for i, req_id in enumerate(  # type: ignore[assignment]
-                self.input_batch.req_ids[:num_reqs]):
+        req_ids: List[str] = []
+        for i, req_id in zip(range(num_reqs), self.input_batch.req_ids):
```
Comment (comaniac) [inline vllm/v1/worker/gpu_model_runner.py:879]: Why not just use enumerate? Isn't the num_reqs not always equal to len(req_ids)?

```suggestion
        for i, req_id in enumerate(self.input_batch.req_ids):
```

**Comment by WoosukKwon** [comment]
@comaniac @njhill @LiuXiaoxuanPKU I've update the PR with some simplification for spec decoding. PTAL.

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:951]: Unfortunately, `req_ids` is a list of size `max_num_reqs` at the moment. I think @njhill has a PR to fix this, but I didn't include it in this PR.

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:379]: To clarify, this comment means that we don't need to copy the "sampling metadata" (e.g., temperature, top_p, etc.) which is constant during the request's entire lifetime.

File: vllm/v1/core/scheduler.py [hunk 3]
```diff
@@ -187,6 +189,15 @@ def schedule(self) -> "SchedulerOutput":
...
             req_index += 1
 
+            # Speculative decode related.
+            if request.spec_token_ids:
+                num_scheduled_spec_tokens = (num_new_tokens +
+                                             request.num_computed_tokens -
+                                             request.num_tokens)
+                if num_scheduled_spec_tokens > 0:
+                    scheduled_spec_decode_tokens[request.request_id] = (
+                        request.spec_token_ids[:num_scheduled_spec_tokens])
```
Comment (WoosukKwon) [inline vllm/v1/core/scheduler.py:199]: This part fixes a bug in #12193

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -438,78 +456,6 @@ def _prepare_inputs(
...
-                num_spec_tokens_list) + spec_offsets
-            # cumsums_spec_offsets: [0, 0, 0, 2M, 2M, 4M] + [2, 3, 4, 4, 5, 3]
-            cumsums_spec_offsets = (
-                spec_seq_offsets +
-                spec_req_indices * self.input_batch.token_ids_cpu.shape[1])
-            cumsums_spec_offsets = torch.from_numpy(cumsums_spec_offsets).to(
-                torch.int64)
-            all_spec_token_ids = torch.tensor(all_spec_token_ids,
-                                              device="cpu",
-                                              dtype=self.input_ids_cpu.dtype)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:474]: This part can be skipped as we insert `spec_token_ids` into `token_ids_cpu` and treat them as regular input tokens.

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -438,78 +456,6 @@ def _prepare_inputs(
...
-            # Step 3.  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
-            #       -  [0, 0, 0, 0, 4, 5, 5, 5, 8, 9, 9]
-            #      -> [0, 1, 2, 3, 0, 0, 1, 2, 0, 0, 1]
-            total_num_sampled_tokens = num_sampled_tokens.sum()
-            sampled_arange = (self.arange_np[:total_num_sampled_tokens] -
-                              cumsums_sampled_offsets)
-
-            # [0, 0, 0, 0, 103, 104, 104, 104, 206, 207, 207] ->
-            # [0, 1, 2, 3, 103, 104, 105, 106, 206, 207, 208]
-            spec_decode_logits_indices = logits_start_loc + sampled_arange
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:511]: This part was moved to a separate method for better readability.

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (comaniac) [reply vllm/v1/worker/gpu_model_runner.py:951]: Might be better to comment it, as using zip to truncate a list for a loop is a bit tricky

Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:965]: sg. Added in [c833429](https://github.com/vllm-project/vllm/pull/13339/commits/c833429ceae6748a05a9a4c06bd226e074f22843)

File: vllm/v1/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -958,28 +958,24 @@ def execute_model(
...
                 generator = self.input_batch.generators.get(i)
                 if generator is not None:
                     # This relies on cuda-specific torch-internal impl details
                     generator.set_offset(generator.get_offset() - 4)
 
-        # num_reqs entries should be non-None
-        assert all(
-            req_id is not None for req_id in
-            self.input_batch.req_ids[:num_reqs]), "req_ids contains None"
-        req_ids = cast(List[str], self.input_batch.req_ids[:num_reqs])
```
Comment (ywang96) [inline vllm/v1/worker/gpu_model_runner.py:981]: I thought casting will be much faster? Is there a reason why you did the loop & append instead?

**Review by ywang96 — COMMENTED** [review]
Overall logic looks good to me but I left two comments - PTAL!

File: vllm/v1/worker/gpu_model_runner.py (hunk 6 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:981]: Agree.

Also since it's a list I think it's probably better to not preallocate to the max size like we do for tensors. The list can grow/shrink as needed. So IMO we could change that in the input batch (I actually did that in https://github.com/vllm-project/vllm/pull/13244). This way we can also just keep the type as `List[str]`. (not suggesting for this PR though...)

File: vllm/v1/core/scheduler.py [hunk 7]
```diff
@@ -324,23 +330,24 @@ def schedule(self) -> "SchedulerOutput":
...
-                                        req_to_new_block_ids[req.request_id],
-                                        req.num_computed_tokens)
+                                        req_to_new_block_ids[req.request_id])
             for req in scheduled_new_reqs
         ]
         resumed_reqs_data = [
             self._make_cached_request_data(
                 req,
+                num_scheduled_tokens[req.request_id],
+                scheduled_spec_decode_tokens.get(req.request_id, []),
```
Comment (njhill) [inline vllm/v1/core/scheduler.py:340]: nit: it's a bit cheaper to use `()` instead of `[]` for this kind of thing (but there's probably a bunch of other places we coudl change too)

**Review by njhill — APPROVED** [review]
looks great to me, it's also a nice simplification!

Only minor suggestions

File: vllm/v1/core/scheduler.py (hunk 7 above)
Comment (WoosukKwon) [reply vllm/v1/core/scheduler.py:340]: Oh didn't know that! Fixed!

File: vllm/v1/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -760,6 +708,56 @@ def _calc_mrope_positions(self, scheduler_output: "SchedulerOutput"):
...
+        # Get the number of spec decode tokens for each request.
+        num_reqs = self.input_batch.num_reqs
+        num_spec_decode_tokens_list: List[int] = []
+        for req_id in self.input_batch.req_ids[:num_reqs]:
+            assert req_id is not None
+            spec_token_ids = scheduler_output.scheduled_spec_decode_tokens.get(
+                req_id, [])
+            num_spec_decode_tokens_list.append(len(spec_token_ids))
+        num_spec_decode_tokens = np.array(num_spec_decode_tokens_list,
+                                          dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:725]: Thanks for the suggestion. Fixed.

File: vllm/v1/worker/gpu_model_runner.py [hunk 9]
```diff
@@ -386,23 +414,13 @@ def _prepare_inputs(
...
-                req_id, [])
-            all_spec_token_ids.extend(spec_token_ids)
-            num_spec_tokens_list.append(len(spec_token_ids))
-
-        num_scheduled_tokens: np.ndarray = np.array(num_scheduled_tokens_list,
-                                                    dtype=np.int32)
-        assert max_num_scheduled_tokens > 0
+        num_scheduled_tokens = np.array(num_scheduled_tokens_list,
+                                        dtype=np.int32)
+        max_num_scheduled_tokens = num_scheduled_tokens.max()
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py]: Nice simplification :) Fixed!

File: vllm/v1/core/scheduler.py [hunk 10]
```diff
@@ -366,22 +373,29 @@ def schedule(self) -> "SchedulerOutput":
     def _make_cached_request_data(
         self,
         request: Request,
+        num_scheduled_tokens: int,
+        scheduled_spec_token_ids: List[int],
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/core/scheduler.py]: Nice catch. Fixed!

File: vllm/v1/worker/gpu_model_runner.py (hunk 6 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:983]: Regarding this particular line, I think the proposed change would be slightly faster because the original code creates the list three times (two from `req_ids[:num_reqs]` and another from the list comprehension for `all`) while the proposed change creates only one.

Agreed with @njhill. All of these are quite hacky and unnecessarily complex. `req_ids` should be fixed by #13244.
